Spot light usd lux lights

ericmehl opened the pull request:
This adds support for USDLux lights to the light tool.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

johnhaddon commented on python/GafferSceneUITest/ParameterInspectorTest.py:
```diff
@@ -558,12 +565,47 @@ def testNonExistentLocation( self ) :
+		editScope["in"].setInput( light["out"] )
+
 		self.assertIsNone( self.__inspect( light["out"], "/light", "exposure", attribute = "nothingHere" ) )
+		self.assertIsNone( self.__inspect( light["out"], "/light", "exposure", editScope, attribute = "nothingHere" ) )
```
Should this be `__inspect( editScope["out"], ... )`?

johnhaddon commented on python/GafferSceneUITest/ParameterInspectorTest.py:
```diff
@@ -558,12 +565,47 @@ def testNonExistentLocation( self ) :
+		editScope["in"].setInput( light["out"] )
+
 		self.assertIsNone( self.__inspect( light["out"], "/light", "nothingHere" ) )
+		self.assertIsNone( self.__inspect( light["out"], "/light", "nothingHere", editScope ) )
```
Should this be `__inspect( editScope["out"], ... )` as well?

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5416/commits/47f0781834fb8e5d80c2a458574ba1a92d3d9a28

johnhaddon commented on src/GafferSceneUI/ParameterInspector.cpp:
```diff
@@ -167,6 +172,39 @@ Inspector::EditFunctionOrFailure ParameterInspector::editFunction( Gaffer::EditS
 {
 	auto attributeHistory = static_cast<const SceneAlgo::AttributeHistory *>( history );
 
+	/// \todo This duplicates the logic from `::parameterValue` in `EditScopeAlgo`. Should that
+	/// be exposed publicly in some way?
+
+	auto shaderNetwork = runTimeCast<const ShaderNetwork>( attributeHistory->attributeValue );
+	if( !shaderNetwork )
+	{
+		return fmt::format( "Attribute \"{}\" is not a shader.", attributeHistory->attributeName.string() );
+	}
+
+	const IECoreScene::Shader *shader;
+	if( m_parameter.shader.string().size() )
+	{
+		shader = shaderNetwork->getShader( m_parameter.shader );
+		if( !shader )
+		{
+			return fmt::format( "Shader \"{}\" does not exist.", m_parameter.shader.string() );
+		}
+	}
+	else
+	{
+		shader = shaderNetwork->outputShader();
+		if( !shader )
+		{
+			return "Output shader does not exist.";
+		}
+	}
+
+	const Data *result = shader->parametersData()->member( m_parameter.name );
+	if( !result )
+	{
+		return fmt::format( "Parameter \"{}\" does not exist.", m_parameter.name.string() );
+	}
```
Rather than duplicate all this, I'd be very tempted to just call `ParameterInspector::value()` and use a generic "Parameter does not exist" message for null. Longer term I think we probably want to have some sort of registry of default values, so that we can create parameters from out of the blue, like we do for some attributes.

johnhaddon reviewed (commented):
Thanks for the quick turnaround on this Eric - I've made a couple of comments inline, but nothing major.

ericmehl commented on Changes.md:
```diff
@@ -12,7 +12,9 @@ Improvements
-- LightTool : Added support for editing animated plugs.
+- LightTool
+  - Added support for editing animated plugs.
+  - Added support for eidting lights with parameters set by `OptionalValuePlug`, such asUSD lights with the `shaping:cone:angle` parameter enabled.
```
Fixed in the commits just now.

ericmehl commented on Changes.md:
```diff
@@ -22,7 +24,10 @@ Fixes
-- LightTool : Fixed bug causing non-settable plugs to be enabled, such as plugs with an expression as input or those using the default spreadsheet row for their value. This generated the error `ERROR : EventSignalCombiner : Cannot set value for plug {plug} except during computation.`
+- LightTool
+  - Fixed bug causing non-settable plugs to be enabled, such as plugs with an expression as input or those using the default spreadsheet row for their value. This generated the error `ERROR : EventSignalCombiner : Cannot set value for plug {plug} except during computation.`
+  - Fixed crash that would occur when enabling the light tool with a USD light with `shaping:cone:angle` disabled.
```
Fixed in the commits just now.

ericmehl commented:
We should be good on CI now, it passed a minute ago but I noticed a mistake `changes.md` so I've pushed a fix for that. I've addressed the comments so as long as nothing weird happens to CI this should be ready for another look and hopefully a quick squash and merge.
